lib/finch/conn.ex

```diff
@@ -9,22 +9,26 @@ defmodule Finch.Conn do
       scheme: scheme,
       host: host,
       port: port,
-      opts: opts,
+      opts: opts.conn_opts,
       parent: parent,
+      backoff: opts.backoff,
       mint: nil
     }
   end
 
-  def connect(%{mint: mint} = conn) when not is_nil(mint), do: conn
+  def connect(conn), do: connect(conn, conn.backoff.initial)
 
-  def connect(conn) do
-    # TODO add back-off
+  def connect(%{mint: mint} = conn, _) when not is_nil(mint), do: conn
+
+  def connect(conn, prev_backoff) do
     with {:ok, mint} <- HTTP.connect(conn.scheme, conn.host, conn.port, conn.opts),
          {:ok, mint} <- HTTP.controlling_process(mint, conn.parent) do
       %{conn | mint: mint}
     else
       _ ->
-        conn
+        backoff = :backoff.rand_increment(prev_backoff, conn.backoff.max)
+        Process.sleep(backoff)
```

Comment:
Ideally you wouldn't sleep but rather send a message to yourself but we don't have a receive loop here, so the best option is to indeed to sleep for now. :)